rfcs/0000_intelligent_resampling.md

```diff
@@ -0,0 +1,368 @@
+
+For Case 1b, we need a representation of "null" observations to compute prediction errors. We define "null" features as the absence of **morphological** features (surface normal and principal curvatures), since depending on the sensing modality (vision or touch), not all **non-morphological features** may be detected.
+
+**Current representation**: In the existing implementation, off-object observations result in an empty dictionary for `morphological_features`:
+
+```python
+# relevant code from sensor_modules.py
+if obs_3d[center_id][3] or (
+    not on_object_only and features["object_coverage"] > 0
+):
+    (
+        features,
+        morphological_features,
+        invalid_signals,
+    ) = self.extract_and_add_features(
+        features,
+        obs_3d,
+        rgba_feat,
+        depth_feat,
+        center_id,
+        center_row_col,
+        sensor_frame_data,
+        world_camera,
+    )
+else:
+    invalid_signals = True
+    morphological_features = {}
+
+...
+
+observed_state = State(
+    location=np.array([x, y, z]),
+    morphological_features=morphological_features,
+    non_morphological_features=features, # Note: This may throw an error since 'features' is undefined in off-object cases
+    confidence=1.0,
+    use_state=bool(morphological_features["on_object"]) and not invalid_signals,
+    sender_id=self.sensor_module_id,
+    sender_type="SM",
+)
+```
+
+**Proposed representation**: To enable meaningful prediction error calculations, we may use some default null values that maintain consistent data structure:
+
+```python
+NULL_MORPHOLOGICAL_FEATURES = {
+    "pose_vectors": np.array([np.nan, np.nan, np.nan]),
+    "principal_curvatures": np.array([np.nan, np.nan]),
+    "mean_curvature": np.nan,
+    "gaussian_curvature": np.nan,
+    "on_object": False,
+}
+```
+This representation ensures that all observations have consistent keys regardless of on/off-object status.
+
+**Note**: Off-object observations should trigger FeatureChangeSM but **should not create new nodes in the object model**
+
+### Case 2: Hypothesis remains within the object but expected features are incorrect due to location or pose mismatch
+
+<img src="./0000_intelligent_resampling/prediction_error_case2.png" alt="Prediction Error Case 2" style="width:80%; height:auto; display: block; margin: 0 auto;"/>
+
+_Figure 5_. Hypothesis remains within the object boundaries but at an incorrect location. The hypothesis believes it moved from the mug body to the handle, but actually moved to a different location on the body. This location mismatch results in incorrect feature predictions.
+
+This scenario is already handled by the existing `hypotheses_displacer.py`, which computes evidence based on feature mismatches. No changes are needed for Case 2, except to handle null observations from Case 1b.
```

Comment:
Why would we get null observations in this case? As far as I can see Case 2 is irrelevant to this question as it is already completely handled by the current algorithm.